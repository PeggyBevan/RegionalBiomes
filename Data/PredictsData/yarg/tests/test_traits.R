TestTraitNumeric <- function() {
  res <- TraitDataset('Test', 'Trait 1', 'x', taxa=c('A','B'), kingdoms='Animalia', ranks='Species', values=1:2)
  AssertEqual(res$Reference,'Test')
  AssertEqual(res$Trait, 'Trait 1')
  AssertEqual(res$Unit, 'x')
  expected.values <- data.frame(Taxon=c('A','B'), Rank='Species', Kingdom='Animalia', Value=1:2)
  # Factor levels for rank
  expected.values$Rank <- EnsureFactorLevels(expected.values$Rank, TaxonomicRanks())
  AssertEqual(res$Values, expected.values)
}

TestTraitCategorical <- function() {
  res <- TraitDataset('Test', 'Trait 1', 'x', taxa=c('A','B'), kingdoms='Animalia', ranks='Species', values=c('CARNIVORE', 'HERBIVORE'))
  AssertEqual(res$Reference,'Test')
  AssertEqual(res$Trait, 'Trait 1')
  AssertEqual(res$Unit, 'x')
  expected.values <- data.frame(Taxon=c('A','B'), Rank='Species', Kingdom='Animalia', Value=c('CARNIVORE', 'HERBIVORE'))
  # Factor levels for rank
  expected.values$Rank <- EnsureFactorLevels(expected.values$Rank, TaxonomicRanks())
  AssertEqual(res$Values, expected.values)
}

TestTraitLogical <- function() {
  res <- TraitDataset('Test', 'Trait 1', 'x', taxa=c('A','B'), kingdoms='Animalia', ranks='Species', values=c(TRUE, FALSE))
  AssertEqual(res$Reference,'Test')
  AssertEqual(res$Trait, 'Trait 1')
  AssertEqual(res$Unit, 'x')
  expected.values <- data.frame(Taxon=c('A','B'), Rank='Species', Kingdom='Animalia', Value=c(TRUE, FALSE))
  # Factor levels for rank
  expected.values$Rank <- EnsureFactorLevels(expected.values$Rank, TaxonomicRanks())
  AssertEqual(res$Values, expected.values)
}

TestTraitFailures <- function() {
  # Empty Reference
  AssertRaises(TraitDataset('',     'Trait 1', 'x', taxa=c('A','B'), kingdoms='Animalia', ranks='Species', values=1:2))

  # Empty Trait
  AssertRaises(TraitDataset('Test', '',        'x', taxa=c('A','B'), kingdoms='Animalia', ranks='Species', values=1:2))

  # Empty Unit
  AssertRaises(TraitDataset('Test', 'Trait 1', '',  taxa=c('A','B'), kingdoms='Animalia', ranks='Species', values=1:2))

  # Duplicated taxa
  AssertRaises(TraitDataset('Test', 'Trait 1', 'x', taxa=c('A','A'), kingdoms='Animalia', ranks='Species', values=1:2))

  # Unknown rank
  AssertRaises(TraitDataset('Test', 'Trait 1', 'x', taxa=c('A','B'), kingdoms='Animalia', ranks='x',       values=1:2))
}

TestAddTraits <- function() {
  input <- read.csv('test_traits_input.csv')
  input$Genus <- sapply(strsplit(as.character(input$Taxon),  ' '), '[', 1)
  traits <- read.csv('test_traits_numeric.csv')
  numeric1 <- with(traits['Numeric 1'==traits$Trait,], TraitDataset('Test','Numeric 1','x',Taxon,'Species',Kingdom,Value))
  numeric2 <- with(traits['Numeric 2'==traits$Trait,], TraitDataset('Test','Numeric 2','x',Taxon,'Species',Kingdom,Value))
  numeric3 <- with(traits['Numeric 3'==traits$Trait,], TraitDataset('Test','Numeric 3','x',Taxon,'Species',Kingdom,Value))

  traits <- read.csv('test_traits_categorical.csv')
  categorical1 <- TraitDataset('Test','Categorical 1','x',traits$Taxon,'Species',traits$Kingdom,traits$Value)

  traits <- read.csv('test_traits_logical.csv')
  logical1 <- TraitDataset('Test','Logical 1','x',traits$Taxon,'Species',traits$Kingdom,traits$Value)

  res <- input
  res <- AddTraits(res, numeric1, numeric2, numeric3, categorical1, logical1)
  res <- AddGenusAveragedTraits(res, numeric1, numeric2, numeric3)

  # Can't genus-averaged categorical or logical traits
  AssertRaises(AddGenusAveragedTraits(res, categorical1))
  AssertRaises(AddGenusAveragedTraits(res, logical1))

  expected <- read.csv('test_traits_expected.csv', check.names=FALSE)

  res$Categorical_1_x <- as.character(res$Categorical_1_x)
  expected$Categorical_1_x <- as.character(expected$Categorical_1_x)
  AssertEqual(expected[,intersect(colnames(res), colnames(expected))], 
              res[,intersect(colnames(res), colnames(expected))])

  # Existing trait values should not be altered
  res <- AddTraits(res, numeric1, numeric2, numeric3, categorical1)
  res <- AddGenusAveragedTraits(res, numeric1, numeric2, numeric3)
  AssertEqual(expected[,intersect(colnames(res), colnames(expected))], 
              res[,intersect(colnames(res), colnames(expected))])

  cat('TODO test InferTraitValues\n')
}
