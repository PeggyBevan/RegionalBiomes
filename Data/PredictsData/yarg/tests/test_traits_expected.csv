Taxon,Kingdom,Numeric_1_x,Numeric_2_x,Numeric_3_x,Numeric 1 notes,Numeric 2 notes,Numeric 3 notes,Categorical_1_x,Logical_1_x
A a,Animalia,1,100,NA,Exact,Exact,,Predator,TRUE
A b,Animalia,2,105,NA,Exact,Mean of all A,,Predator
A c,Animalia,3,105,NA,Exact,Mean of all A,,Predator
A d,Animalia,4,105,NA,Exact,Mean of all A,,Predator
A e,Animalia,5,110,NA,Exact,Exact,,Predator,FALSE
A e,Animalia,5,110,NA,Exact,Exact,,Predator,FALSE
A e,Animalia,5,110,NA,Exact,Exact,,Predator,FALSE
A a,Plantae,NA,NA,0.1,,,Exact,Producer
A e,Plantae,NA,NA,0.5,,,Exact,Producer
A f,Plantae,NA,NA,0.3,,,Mean of all A,Producer
A a,Animalia,1,100,NA,Exact,Exact,,Predator,TRUE
A,Animalia,3,105,NA,Mean of all A,Mean of all A,,NA
A,Animalia,3,105,NA,Mean of all A,Mean of all A,,NA
A f,Animalia,3,105,NA,Mean of all A,Mean of all A,,NA
A g,Animalia,3,105,NA,Mean of all A,Mean of all A,,NA
D,Animalia,NA,NA,NA,,,,NA
E,Animalia,NA,NA,NA,,,,NA
B a,Animalia,15,120,NA,Exact,Exact,,Parasite
B b,Animalia,25,130,NA,Exact,Exact,,Parasite
B c,Animalia,35,130,NA,Exact,Mean of all B,,NA
B d,Animalia,45,130,NA,Exact,Mean of all B,,NA
B e,Animalia,55,140,NA,Exact,Mean of all B,,NA
B f,Animalia,35,130,NA,Mean of all B,Mean of all B,,NA
B,Animalia,35,130,NA,Mean of all B,Mean of all B,,NA
B,Animalia,35,130,NA,Mean of all B,Mean of all B,,NA
